File: backend/app/dm/review/schemas.py
"""簽核中心（US6 / DM04）schema。

待簽核清單 / 明細（新舊版並列）/ 核准並發布 / 退回 / 已完成。回應對齊 spec_us6 FR-001~008。
"""

from datetime import datetime

from pydantic import BaseModel


class PendingItem(BaseModel):
    """待簽核清單列（指派給當前審核者之 PENDING）。"""

    review_id: int
    doc_id: str
    doc_name: str
    category_code: str
    review_type: str  # NEW / NEW_VERSION / OBSOLETE
    version_no: str | None  # 送審版本號（草稿版可能為空）
    submitter_id: str  # 送審者（review.created_user）
    submitter_name: str | None
    submit_date: datetime
    waiting_days: int  # 停留天數（送審至今）


class VersionMeta(BaseModel):
    """版本檔案 metadata（明細之新 / 舊版比對用）。"""

    version_id: int
    version_no: str | None
    file_name: str | None
    file_size: int | None
    file_mime: str | None
    previewable: bool


class ReviewDetail(BaseModel):
    """簽核明細：變更摘要 + 送審檔案（新版本另附目前發布版供比對）。"""

    review_id: int
    doc_id: str
    doc_name: str
    category_code: str
    review_type: str
    change_summary: str | None
    submit_date: datetime
    submitter_id: str
    submitter_name: str | None
    new_version: VersionMeta | None  # 送審版本
    current_version: VersionMeta | None  # 目前發布版（NEW_VERSION 比對；NEW 無）


class RejectReq(BaseModel):
    """退回請求（JSON）。"""

    reason: str


class ApproveResult(BaseModel):
    """核准並發布結果。"""

    published_version_id: int
    notified: int  # 排入 Email 之收件數（0=範本停用 / 無收件）


class RejectResult(BaseModel):
    """退回結果。"""

    review_id: int


class CompletedItem(BaseModel):
    """已完成清單列（當前審核者過往處理）。"""

    review_id: int
    doc_id: str
    doc_name: str
    review_type: str
    status: str  # APPROVED / REJECTED
    version_no: str | None
    complete_date: datetime | None

File: frontend/src/dm/review/schemas.ts
/** 簽核處理（US6 / DM04）型別與表單驗證（對齊後端 app/dm/review/schemas.py）。 */

import { z } from "zod"

/** 退回原因表單驗證（必填非空；後端另有 max_length 500）。 */
export const RejectReqSchema = z.object({
  reason: z.string().trim().min(1, { message: "請填寫退回原因" }),
})

export interface PendingItem {
  review_id: number
  doc_id: string
  doc_name: string
  category_code: string
  review_type: string // NEW / NEW_VERSION / OBSOLETE
  version_no: string | null
  submitter_id: string
  submitter_name: string | null
  submit_date: string
  waiting_days: number
}

export interface VersionMeta {
  version_id: number
  version_no: string | null
  file_name: string | null
  file_size: number | null
  file_mime: string | null
  previewable: boolean
}

export interface ReviewDetail {
  review_id: number
  doc_id: string
  doc_name: string
  category_code: string
  review_type: string
  change_summary: string | null
  submit_date: string
  submitter_id: string
  submitter_name: string | null
  new_version: VersionMeta | null
  current_version: VersionMeta | null // 新版本申請附目前發布版供比對；首版為 null
  obsolete_reason: string | null // 廢止原因（OBSOLETE）
  obsolete_file_name: string | null // 廢止附件檔名（OBSOLETE；有則可下載）
  obsolete_file_size: number | null // 廢止附件大小（位元組）
}

export interface CompletedItem {
  review_id: number
  doc_id: string
  doc_name: string
  review_type: string
  status: string // APPROVED / REJECTED
  version_no: string | null
  complete_date: string | null
}

export interface ApproveResult {
  published_version_id: number
  notified: number
}

export interface RejectResult {
  review_id: number
}

/** 送審類型顯示名。 */
export const REVIEW_TYPE_LABELS: Record<string, string> = {
  NEW: "新增",
  NEW_VERSION: "新版本",
  OBSOLETE: "廢止",
}

/** 停留天數標紅門檻（與後端 DM_REMIND_THRESHOLD 預設一致；逾此於清單標紅警示）。 */
export const REMIND_THRESHOLD_DAYS = 7

/** 已完成狀態顯示名。 */
export const REVIEW_STATUS_LABELS: Record<string, string> = {
  APPROVED: "已核准",
  REJECTED: "已退回",
}
